src/mlpack/methods/isomap/k_nearest.hpp

```diff
@@ -0,0 +1,115 @@
+/**
+ * @file k_nearest.hpp
+ * @author Rishabh Ranjan
+ * 
+ * This file defines the class K_Nearest which constructs the neighbourhood
+ * graph for Isomap, according to the number of neighbours provided.
+ * 
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+
+#ifndef MLPACK_METHODS_ISOMAP_K_NEAREST
+#define MLPACK_METHODS_ISOMAP_K_NEAREST
+
+#include <mlpack/prereqs.hpp>
+#include <queue>
+
+namespace mlpack {
+namespace isomap {
+/**
+ * This class implements the K-Nearest Neighbours function to construct
+ * a neighbourhood graph for the given distance matrix.
+*/
+
+class K_Nearest
+{
+ public:
+  /**
+   * Function to make the neighbourhood graph for the given distance matrix
+   * and number of neighbours.
+   * 
+   * @param disMat -distance matrix to construct the graph
+   * @param n_neighbours -number of neighbours to consider
+   */
+  void MakeNeighbourhoodGraph(arma::mat &disMat,
+                              const int n_neighbours)
+  {
+    // go through every row
+    for (size_t i = 0; i < disMat.n_rows; i++)
```

Comment:
Armadillo stores matrices in column-major format, so it can be orders of magnitude faster to iterate over columns instead of iterating over rows.  To be more specific, this means that in memory, points in a single column are stored contiguously.

So, whereever you are iterating over matrices, I would advise you to do it column-by-column instead, even if this means storing the matrix in what seems like a transposed form.  It can have major speed benefits to do that.

Reply:
Oh, nice. I wasn't aware of that. Next commit will resolve this.